// typescript-json/lib/metadata/MetadataObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataObject = void 0;
var MetadataObject = (function () {
    function MetadataObject(props) {
        this.nullables = [];
        this.name = props.name;
        this.properties = props.properties;
        this.description = props.description;
        this.jsDocTags = props.jsDocTags;
        this.index = props.index;
        this.validated = props.validated;
        this.recursive = false;
        this.nullables = [];
    }
    MetadataObject.create = function (props) {
        return new MetadataObject(props);
    };
    MetadataObject._From_without_properties = function (obj) {
        return this.create({
            name: obj.name,
            properties: [],
            description: obj.description,
            jsDocTags: obj.jsDocTags,
            index: obj.index,
            validated: obj.validated,
            recursive: obj.recursive,
            nullables: obj.nullables.slice(),
        });
    };
    MetadataObject.prototype.toJSON = function () {
        return {
            name: this.name,
            properties: this.properties.map(function (property) { return property.toJSON(); }),
            description: this.description,
            jsDocTags: this.jsDocTags,
            index: this.index,
            validated: this.validated,
            recursive: this.recursive,
            nullables: this.nullables.slice(),
        };
    };
    return MetadataObject;
}());
exports.MetadataObject = MetadataObject;
(function (MetadataObject) {
    function intersects(x, y) {
        return x.properties.some(function (prop) {
            return y.properties.find(function (oppo) { return prop.key === oppo.key; }) !==
                undefined;
        });
    }
    MetadataObject.intersects = intersects;
    function covers(x, y) {
        return (x.properties.length >= y.properties.length &&
            x.properties.every(function (prop) {
                return y.properties.find(function (oppo) { return prop.key === oppo.key; }) !==
                    undefined;
            }));
    }
    MetadataObject.covers = covers;
})(MetadataObject = exports.MetadataObject || (exports.MetadataObject = {}));
exports.MetadataObject = MetadataObject;
