// typescript-json/lib/programmers/internal/application_object.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.application_object = void 0;
var PatternUtil_1 = require("../../utils/PatternUtil");
var application_schema_1 = require("./application_schema");
var metadata_to_pattern_1 = require("./metadata_to_pattern");
var application_object = function (options) {
    return function (components) {
        return function (key, obj, nullable) {
            var e_1, _a;
            if (components.schemas[key] !== undefined)
                return;
            components.schemas[key] = {};
            var properties = {};
            var patternProperties = {};
            var additionalProperties = [];
            var required = [];
            try {
                for (var _b = __values(obj.properties), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var property = _c.value;
                    if (property.value.functional === true &&
                        property.value.nullable === false &&
                        property.value.required === true &&
                        property.value.size() === 0)
                        continue;
                    var key_1 = property.key.getSoleLiteral();
                    var value = (0, application_schema_1.application_schema)(options)(components)(true)(property.value, {
                        description: property.description,
                        "x-tson-metaTags": property.tags.length
                            ? property.tags
                            : undefined,
                        "x-tson-jsDocTags": property.jsDocTags.length
                            ? property.jsDocTags
                            : undefined,
                    });
                    if (value === null)
                        continue;
                    else if (key_1 !== null) {
                        properties[key_1] = value;
                        if (property.value.required === true)
                            required.push(key_1);
                    }
                    else {
                        var pattern = (0, metadata_to_pattern_1.metadata_to_pattern)(true)(property.key);
                        if (options.purpose === "swagger" ||
                            pattern === PatternUtil_1.PatternUtil.STRING)
                            additionalProperties.push(value);
                        else
                            patternProperties[pattern] = value;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            var schema = {
                $id: options.purpose === "ajv"
                    ? options.prefix + "/" + key
                    : undefined,
                $recursiveAnchor: (options.purpose === "ajv" && obj.recursive) || undefined,
                type: "object",
                properties: properties,
                patternProperties: Object.keys(patternProperties).length
                    ? patternProperties
                    : undefined,
                additionalProperties: additionalProperties.length
                    ? additionalProperties.length === 1
                        ? additionalProperties[0]
                        : { oneOf: additionalProperties }
                    : undefined,
                nullable: nullable,
                required: required.length ? required : undefined,
                description: obj.description,
                "x-tson_jsDocTags": obj.jsDocTags,
            };
            components.schemas[key] = schema;
        };
    };
};
exports.application_object = application_object;
