Index: udan-react-ts/build/manifest.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.BaseRevisionTextPatchEP
<+>{\r\n  \"manifest_version\": 3,\r\n  \"name\": \"React Chrome Injection : Lak\",\r\n  \"version\": \"0.0.1\",\r\n  \"content_scripts\": [\r\n    {\r\n      \"matches\": [\r\n        \"http://*/*\",\r\n        \"https://*/*\"\r\n      ],\r\n      \"run_at\": \"document_end\",\r\n      \"js\": [\r\n        \"index.js\"\r\n      ],\r\n      \"css\": [\r\n        \"assets/App.min.css\"\r\n      ],\r\n      \"media\": []\r\n    }\r\n  ],\r\n  \"background\": {\r\n    \"service_worker\": \"./background.js\"\r\n  },\r\n  \"permissions\": [\r\n    \"activeTab\",\r\n    \"identity\",\r\n    \"identity.email\",\r\n    \"storage\"\r\n  ],\r\n  \"web_accessible_resources\": [\r\n    {\r\n      \"resources\": [\r\n        \"assets/bundle.js\",\r\n        \"index.js\",\r\n        \"assets/App.min.css\",\r\n        \"assets/logo.svg\"\r\n      ],\r\n      \"matches\": [\r\n        \"http://*/*\",\r\n        \"https://*/*\"\r\n      ]\r\n    }\r\n  ]\r\n}
===================================================================
diff --git a/udan-react-ts/build/manifest.json b/udan-react-ts/build/manifest.json
--- a/udan-react-ts/build/manifest.json	
+++ b/udan-react-ts/build/manifest.json	
@@ -1,6 +1,6 @@
 {
   "manifest_version": 3,
-  "name": "React Chrome Injection : Lak",
+  "name": "UDAN-React-Library",
   "version": "0.0.1",
   "content_scripts": [
     {
@@ -41,4 +41,4 @@
       ]
     }
   ]
-}
\ No newline at end of file
+}
