// typescript-json/lib/programmers/AssertProgrammer.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssertProgrammer = void 0;
var typescript_1 = __importDefault(require("typescript"));
var IdentifierFactory_1 = require("../factories/IdentifierFactory");
var ValueFactory_1 = require("../factories/ValueFactory");
var CheckerProgrammer_1 = require("./CheckerProgrammer");
var IsProgrammer_1 = require("./IsProgrammer");
var FunctionImporeter_1 = require("./helpers/FunctionImporeter");
var check_object_1 = require("./internal/check_object");
var AssertProgrammer;
(function (AssertProgrammer) {
    AssertProgrammer.generate = function (project, modulo, equals) {
        if (equals === void 0) { equals = false; }
        return function (type) {
            var importer = new FunctionImporeter_1.FunctionImporter();
            var program = CheckerProgrammer_1.CheckerProgrammer.generate(project, {
                functors: "$ao",
                unioners: "$au",
                path: true,
                trace: true,
                numeric: !!project.options.numeric,
                equals: equals,
                combiner: combiner(equals)(importer),
                joiner: joiner(equals)(importer),
                success: typescript_1.default.factory.createTrue(),
            }, importer)(type);
            return typescript_1.default.factory.createArrowFunction(undefined, undefined, [IdentifierFactory_1.IdentifierFactory.parameter("input")], undefined, undefined, typescript_1.default.factory.createBlock(__spreadArray(__spreadArray([], __read(importer.declare(modulo)), false), [
                typescript_1.default.factory.createExpressionStatement(typescript_1.default.factory.createCallExpression(program, undefined, [
                    ValueFactory_1.ValueFactory.INPUT(),
                    typescript_1.default.factory.createStringLiteral("$input"),
                    typescript_1.default.factory.createTrue(),
                ])),
                typescript_1.default.factory.createReturnStatement(ValueFactory_1.ValueFactory.INPUT()),
            ], false), true));
        };
    };
    var combiner = function (equals) {
        return function (importer) {
            return function (explore) {
                if (explore.tracable === false && explore.from !== "top")
                    return IsProgrammer_1.IsProgrammer.CONFIG({
                        object: assert_object(equals)(importer),
                        numeric: true,
                    }).combiner(explore);
                var path = explore.postfix
                    ? "path + ".concat(explore.postfix)
                    : "path";
                return function (logic) { return function (input, binaries, expected) {
                    return logic === "and"
                        ? binaries
                            .map(function (binary) {
                            return binary.combined
                                ? binary.expression
                                : typescript_1.default.factory.createLogicalOr(binary.expression, create_guard_call(importer)(explore.source === "top"
                                    ? typescript_1.default.factory.createTrue()
                                    : typescript_1.default.factory.createIdentifier("exceptionable"))(typescript_1.default.factory.createIdentifier(path), expected, input));
                        })
                            .reduce(typescript_1.default.factory.createLogicalAnd)
                        : (function () {
                            var addicted = binaries.slice();
                            if (addicted[addicted.length - 1].combined === false) {
                                addicted.push({
                                    combined: true,
                                    expression: create_guard_call(importer)(explore.source === "top"
                                        ? typescript_1.default.factory.createTrue()
                                        : typescript_1.default.factory.createIdentifier("exceptionable"))(typescript_1.default.factory.createIdentifier(path), expected, input),
                                });
                            }
                            return addicted
                                .map(function (b) { return b.expression; })
                                .reduce(typescript_1.default.factory.createLogicalOr);
                        })();
                }; };
            };
        };
    };
    var assert_object = function (equals) { return function (importer) {
        return (0, check_object_1.check_object)({
            equals: equals,
            assert: true,
            reduce: typescript_1.default.factory.createLogicalAnd,
            positive: typescript_1.default.factory.createTrue(),
            superfluous: function (value) {
                return create_guard_call(importer)()(typescript_1.default.factory.createAdd(typescript_1.default.factory.createIdentifier("path"), typescript_1.default.factory.createCallExpression(importer.use("join"), undefined, [typescript_1.default.factory.createIdentifier("key")])), "undefined", value);
            },
            halt: function (expr) {
                return typescript_1.default.factory.createLogicalOr(typescript_1.default.factory.createStrictEquality(typescript_1.default.factory.createFalse(), typescript_1.default.factory.createIdentifier("exceptionable")), expr);
            },
        });
    }; };
    var joiner = function (equals) {
        return function (importer) { return ({
            object: assert_object(equals)(importer),
            array: function (input, arrow) {
                return typescript_1.default.factory.createCallExpression(IdentifierFactory_1.IdentifierFactory.join(input, "every"), undefined, [arrow]);
            },
            failure: function (value, expected, explore) {
                return create_guard_call(importer)((explore === null || explore === void 0 ? void 0 : explore.from) === "top"
                    ? typescript_1.default.factory.createTrue()
                    : typescript_1.default.factory.createIdentifier("exceptionable"))(typescript_1.default.factory.createIdentifier((explore === null || explore === void 0 ? void 0 : explore.postfix) ? "path + ".concat(explore.postfix) : "path"), expected, value);
            },
            full: equals
                ? undefined
                : function (condition) { return function (input, expected, explore) {
                    return typescript_1.default.factory.createLogicalOr(condition, create_guard_call(importer)(explore.from === "top"
                        ? typescript_1.default.factory.createTrue()
                        : typescript_1.default.factory.createIdentifier("exceptionable"))(typescript_1.default.factory.createIdentifier("path"), expected, input));
                }; },
        }); };
    };
    var create_guard_call = function (importer) {
        return function (exceptionable) {
            return function (path, expected, value) {
                return typescript_1.default.factory.createCallExpression(importer.use("guard"), undefined, [
                    exceptionable || typescript_1.default.factory.createIdentifier("exceptionable"),
                    typescript_1.default.factory.createObjectLiteralExpression([
                        typescript_1.default.factory.createPropertyAssignment("path", path),
                        typescript_1.default.factory.createPropertyAssignment("expected", typescript_1.default.factory.createStringLiteral(expected)),
                        typescript_1.default.factory.createPropertyAssignment("value", value),
                    ], true),
                ]);
            };
        };
    };
})(AssertProgrammer = exports.AssertProgrammer || (exports.AssertProgrammer = {}));
